README
CMPH - C Minimal Perfect Hashing Library


-------------------------------------------------------------------


	Motivation
	==========

A perfect hash function maps a static set of n keys into a set of m integer numbers without collisions, where m is greater than or equal to n. If m is equal to n, the function is called minimal.

Minimal perfect hash functions (concepts.html) are widely used for memory efficient storage and fast retrieval of items from static sets, such as words in natural languages, reserved words in programming languages or interactive systems, universal resource locations (URLs) in Web search engines, or item sets in data mining techniques. Therefore, there are applications for minimal perfect hash functions in information retrieval systems, database systems, language translation systems, electronic commerce systems, compilers, operating systems, among others.

The use of minimal perfect hash functions is, until now, restricted to scenarios where the set of keys being hashed is small, because of the limitations of current algorithms. But in many cases, to deal with huge set of keys is crucial. So, this project gives to the free software community an API that will work with sets in the order of billion of keys.

Probably, the most interesting application for minimal perfect hash functions is its use as an indexing structure for databases. The most popular data structure used as an indexing structure in databases is the B+ tree. In fact, the B+ tree is very used for dynamic applications with frequent insertions and deletions of records. However, for applications with sporadic modifications and a huge number of queries the B+ tree is not the best option, because practical deployments of this structure are extremely complex, and perform poorly with very large sets of keys such as those required for the new frontiers database applications (http://acmqueue.com/modules.php?name=Content&pa=showpage&pid=299).

For example, in the information retrieval field, the work with huge collections is a daily task. The simple assignment of ids to web pages of a collection can be a challenging task. While traditional databases simply cannot handle more traffic once the working set of web page urls does not fit in main memory anymore, minimal perfect hash functions can easily scale to hundred of millions of entries, using stock hardware.

As there are lots of applications for minimal perfect hash functions, it is important to implement memory and time efficient algorithms for constructing such functions. The lack of similar libraries in the free software world has been the main motivation to create the C Minimal Perfect Hashing Library (gperf is a bit different (gperf.html), since it was conceived to create very fast perfect hash functions for small sets of keys and CMPH Library was conceived to create minimal perfect hash functions for very large sets of keys). C Minimal Perfect Hashing Library is a portable LGPLed library to generate and to work with very efficient minimal perfect hash functions.

-------------------------------------------------------------------


	Description
	===========

The CMPH Library encapsulates the newest and more efficient algorithms in an easy-to-use, production-quality, fast API. The library was designed to work with big entries that cannot fit in the main memory. It has been used successfully for constructing minimal perfect hash functions for sets with more than 100 million of keys, and we intend to expand this number to the order of billion of keys. Although there is a lack of similar libraries, we can point out some of the distinguishable features of the CMPH Library:

- Fast.
- Space-efficient with main memory usage carefully documented.
- The best modern algorithms are available (or at least scheduled for implementation :-)).
- Works with in-disk key sets through of using the adapter pattern.
- Serialization of hash functions.
- Portable C code (currently works on GNU/Linux and WIN32 and is reported to work in OpenBSD and Solaris).
- Object oriented implementation.
- Easily extensible.
- Well encapsulated API aiming binary compatibility through releases.
- Free Software.

----------------------------------------


	Supported Algorithms
	====================

- CHD Algorithm:
  - It is the fastest algorithm to build PHFs and MPHFs in linear time.
  - It generates the most compact PHFs and MPHFs we know of.
  - It can generate PHFs with a load factor up to 99 %.
  - It can be used to generate t-perfect hash functions. A t-perfect hash function allows at most t collisions in a given bin. It is a well-known fact that modern memories are organized as blocks which constitute transfer unit. Example of such blocks are cache lines for internal memory or sectors for hard disks. Thus, it can be very useful for devices that carry out I/O operations in blocks.   
  - It is a two level scheme. It uses  a first level hash function to split the key set in buckets of average size determined by a parameter b in the range [1,32]. In the second level it uses displacement values to resolve the collisions that have given rise to the buckets.
  - It can generate MPHFs that can be stored in approximately 2.07 bits per key.
  - For a load factor equal to the maximum one that is achieved by the BDZ algorithm (81 %), the resulting PHFs are stored in approximately 1.40 bits per key. 
- BDZ Algorithm:
  - It is very simple and efficient. It outperforms all the ones below.
  - It constructs both PHFs and MPHFs in linear time. 
  - The maximum load factor one can achieve for a PHF is 1/1.23.  
  - It is based on acyclic random 3-graphs. A 3-graph is a generalization of a graph where each edge connects 3 vertices instead of only 2. 
  - The resulting MPHFs are not order preserving.
  - The resulting MPHFs can be stored in only (2 + x)cn bits, where c should be larger than or equal to 1.23 and x is a constant larger than 0 (actually, x = 1/b and b is a parameter that should be larger than 2). For c = 1.23 and b = 8, the resulting functions are stored in approximately 2.6 bits per key.
  - For its maximum load factor (81 %), the resulting PHFs are stored in approximately 1.95 bits per key. 
- BMZ Algorithm:
  - Construct MPHFs in linear time.
  - It is based on cyclic random graphs. This makes it faster than the CHM algorithm.
  - The resulting MPHFs are not order preserving. 
  - The resulting MPHFs are more compact than the ones generated by the CHM algorithm and can be stored in 4cn bytes, where c is in the range [0.93,1.15]. 
- BRZ Algorithm:
  - A very fast external memory based algorithm for constructing minimal perfect hash functions  for sets in the order of billions of keys.
  - It works in linear time. 
  - The resulting MPHFs are not order preserving.
  - The resulting MPHFs can be stored using less than 8.0 bits per key.
- CHM Algorithm:
  - Construct minimal MPHFs in linear time. 
  - It is based on acyclic random graphs
  - The resulting MPHFs are order preserving.
  - The resulting MPHFs are stored in 4cn bytes, where c is greater than 2.
- FCH Algorithm:
  - Construct minimal perfect hash functions that require  less than 4 bits per key to be stored. 
  - The resulting MPHFs are very compact and very efficient at evaluation time
  - The algorithm is only efficient for small sets.
  - It is used as internal algorithm in the BRZ algorithm to efficiently solve larger problems and even so to generate MPHFs that require approximately 4.1 bits per key to be stored. For that, you just need to set the parameters -a to brz and -c to a value larger than or equal to 2.6.

----------------------------------------


	News for version 2.0
	====================

Cleaned up most warnings for the c code.

Experimental C++ interface (--enable-cxxmph) implementing the BDZ algorithm in
a convenient SimpleMPHIndex interface, which serves as the basis
for drop-in replacements for std::unordered_map, sparsehash::sparse_hash_map
and sparsehash::dense_hash_map. Faster lookup time at the expense of insertion
time. See cxxmpph/mph_map.h and cxxmph/mph_index.h for details.


	News for version 1.1
	====================

Fixed a bug in the chd_pc algorithm and reorganized tests.


	News for version 1.0
	====================

This is a bugfix only version, after which a revamp of the cmph code and
algorithms will be done.


	News for version 0.9
	====================

- The CHD algorithm (chd.html), which is an algorithm that can be tuned to generate MPHFs that require approximately 2.07 bits per key to be stored. The algorithm outperforms the BDZ algorithm (bdz.html) and therefore is the fastest one available in the literature for sets that can be treated in internal memory.
- The CHD_PH algorithm (chd.html), which is an algorithm to generate PHFs with load factor up to 99 %. It is actually the CHD algorithm without the ranking step. If we set the load factor to 81 %, which is the maximum that can be obtained with the BDZ algorithm (bdz.html), the resulting functions can be stored in 1.40 bits per key. The space requirement increases with the load factor.
- All reported bugs and suggestions have been corrected and included as well.


	News for version 0.8
	====================

- An algorithm to generate MPHFs that require around 2.6 bits per key to be stored (bdz.html), which is referred to as BDZ algorithm. The algorithm is the fastest one available in the literature for sets that can be treated in internal memory.
- An algorithm to generate PHFs with range m = cn, for c > 1.22 (bdz.html), which is referred to as BDZ_PH algorithm. It is actually the BDZ algorithm without the ranking step. The resulting functions can be stored in 1.95 bits per key for c = 1.23 and are considerably faster than the MPHFs generated by the BDZ algorithm.
- An adapter to support a vector of struct as the source of keys has been added. 
- An API to support the ability of packing a perfect hash function into a preallocated contiguous memory space. The computation of a packed function is still faster and can be easily mmapped.
- The hash functions djb2, fnv and sdbm were removed because they do not use random seeds and therefore are not useful for MPHFs algorithms.
- All reported bugs and suggestions have been corrected and included as well.

News log (newslog.html)

----------------------------------------


	Examples
	========

Using cmph is quite simple. Take a look.

  #include <cmph.h>
  #include <string.h>
  // Create minimal perfect hash function from in-memory vector
  int main(int argc, char **argv)
  { 
  
      // Creating a filled vector
      unsigned int i = 0;
      const char *vector[] = {"aaaaaaaaaa", "bbbbbbbbbb", "cccccccccc", "dddddddddd", "eeeeeeeeee", 
          "ffffffffff", "gggggggggg", "hhhhhhhhhh", "iiiiiiiiii", "jjjjjjjjjj"};
      unsigned int nkeys = 10;
      FILE* mphf_fd = fopen("temp.mph", "w");
      // Source of keys
      cmph_io_adapter_t *source = cmph_io_vector_adapter((char **)vector, nkeys);
  
      //Create minimal perfect hash function using the brz algorithm.
      cmph_config_t *config = cmph_config_new(source);
      cmph_config_set_algo(config, CMPH_BRZ);
      cmph_config_set_mphf_fd(config, mphf_fd);
      cmph_t *hash = cmph_new(config);
      cmph_config_destroy(config);
      cmph_dump(hash, mphf_fd); 
      cmph_destroy(hash);	
      fclose(mphf_fd);
  
      //Find key
      mphf_fd = fopen("temp.mph", "r");
      hash = cmph_load(mphf_fd);
      while (i < nkeys) {
          const char *key = vector[i];
          unsigned int id = cmph_search(hash, key, (cmph_uint32)strlen(key));
          fprintf(stderr, "key:%s -- hash:%u\n", key, id);
          i++;
      }
  
      //Destroy hash
      cmph_destroy(hash);
      cmph_io_vector_adapter_destroy(source);   
      fclose(mphf_fd);
      return 0;
  }

Download vector_adapter_ex1.c (examples/vector_adapter_ex1.c). This example does not work in versions below 0.6. You need to update the sources from GIT to make it work.

-------------------------------

  #include <cmph.h>
  #include <stdio.h>
  #include <string.h>
   // Create minimal perfect hash function from in-disk keys using BDZ algorithm
  int main(int argc, char **argv)
  {   
  	 //Open file with newline separated list of keys
  	FILE * keys_fd = fopen("keys.txt", "r");
  	cmph_t *hash = NULL;
  	if (keys_fd == NULL) 
  	{
  	  fprintf(stderr, "File \"keys.txt\" not found\n");
  	  exit(1);
  	}	
  	// Source of keys
  	cmph_io_adapter_t *source = cmph_io_nlfile_adapter(keys_fd);
  
  	cmph_config_t *config = cmph_config_new(source);
  	cmph_config_set_algo(config, CMPH_BDZ);
  	hash = cmph_new(config);
  	cmph_config_destroy(config);
     
  	//Find key
  	const char *key = "jjjjjjjjjj";
  	unsigned int id = cmph_search(hash, key, (cmph_uint32)strlen(key));
  	fprintf(stderr, "Id:%u\n", id);
  	//Destroy hash
  	cmph_destroy(hash);
  	cmph_io_nlfile_adapter_destroy(source);   
  	fclose(keys_fd);
  	return 0;
  }

Download file_adapter_ex2.c (examples/file_adapter_ex2.c) and keys.txt (examples/keys.txt). This example does not work in versions below 0.8. You need to update the sources from GIT to make it work.

Click here to see more examples (examples.html)

--------------------------------------


	The cmph application
	====================

cmph is the name of both the library and the utility
application that comes with this package. You can use the cmph
application for constructing minimal perfect hash functions from the command line. 
The cmph utility
comes with a number of flags, but it is very simple to create and to query 
minimal perfect hash functions:

   $ # Using the chm algorithm (default one) for constructing a mphf for keys in file keys_file
   $ ./cmph -g keys_file
   $ # Query id of keys in the file keys_query
   $ ./cmph -m keys_file.mph keys_query

The additional options let you set most of the parameters you have
available through the C API. Below you can see the full help message for the 
utility.

  usage: cmph [-v] [-h] [-V] [-k nkeys] [-f hash_function] [-g [-c algorithm_dependent_value][-s seed] ] 
              [-a algorithm] [-M memory_in_MB] [-b algorithm_dependent_value] [-t keys_per_bin] [-d tmp_dir] 
              [-m file.mph] keysfile
  Minimum perfect hashing tool
  
    -h	 print this help message
    -c	 c value determines:
      	  * the number of vertices in the graph for the algorithms BMZ and CHM
      	  * the number of bits per key required in the FCH algorithm
      	  * the load factor in the CHD_PH algorithm
    -a	 algorithm - valid values are
      	  * bmz
      	  * bmz8
      	  * chm
      	  * brz
      	  * fch
      	  * bdz
      	  * bdz_ph
      	  * chd_ph
      	  * chd
    -f	 hash function (may be used multiple times) - valid values are
      	  * jenkins
    -V	 print version number and exit
    -v	 increase verbosity (may be used multiple times)
    -k	 number of keys
    -g	 generation mode
    -s	 random seed
    -m	 minimum perfect hash function file 
    -M	 main memory availability (in MB) used in BRZ algorithm 
    -d	 temporary directory used in BRZ algorithm 
    -b	 the meaning of this parameter depends on the algorithm selected in the -a option:
      	  * For BRZ it is used to make the maximal number of keys in a bucket lower than 256.
      	    In this case its value should be an integer in the range [64,175]. Default is 128.
  
      	  * For BDZ it is used to determine the size of some precomputed rank
      	    information and its value should be an integer in the range [3,10]. Default
      	    is 7. The larger is this value, the more compact are the resulting functions
      	    and the slower are them at evaluation time.
  
      	  * For CHD and CHD_PH it is used to set the average number of keys per bucket
      	    and its value should be an integer in the range [1,32]. Default is 4. The
      	    larger is this value, the slower is the construction of the functions.
      	    This parameter has no effect for other algorithms.
  
    -t	 set the number of keys per bin for a t-perfect hashing function. A t-perfect
      	 hash function allows at most t collisions in a given bin. This parameter applies
      	 only to the CHD and CHD_PH algorithms. Its value should be an integer in the
      	 range [1,128]. Defaul is 1
    keysfile	 line separated file with keys


	Additional Documentation
	========================

FAQ (faq.html)


	Downloads
	=========

Use the project page at sourceforge: http://sf.net/projects/cmph


	License Stuff
	=============

Code is under the LGPL and the MPL 1.1. 

----------------------------------------

Enjoy!

Davi de Castro Reis (davi@users.sourceforge.net)

Djamel Belazzougui (db8192@users.sourceforge.net)

Fabiano Cupertino Botelho (fc_botelho@users.sourceforge.net)

Nivio Ziviani (nivio@dcc.ufmg.br)

Last Updated: Sun Jun  3 04:09:55 2012



ChangeLog
2005-08-08 18:34  fc_botelho

	* INSTALL, examples/Makefile, examples/Makefile.in,
	examples/.deps/file_adapter_ex2.Po,
	examples/.deps/vector_adapter_ex1.Po, src/brz.c: [no log message]

2005-08-07 22:00  fc_botelho

	* src/: brz.c, brz.h, brz_structs.h, cmph.c, cmph.h, main.c:
	temporary directory passed by command line

2005-08-07 20:22  fc_botelho

	* src/brz.c: stable version of BRZ

2005-08-06 22:09  fc_botelho

	* src/bmz.c: no message

2005-08-06 22:02  fc_botelho

	* src/bmz.c: no message

2005-08-06 21:45  fc_botelho

	* src/brz.c: fastest version of BRZ

2005-08-06 17:20  fc_botelho

	* src/: bmz.c, brz.c, main.c: [no log message]

2005-07-29 16:43  fc_botelho

	* src/brz.c: BRZ algorithm is almost stable

2005-07-29 15:29  fc_botelho

	* src/: bmz.c, brz.c, brz_structs.h, cmph_types.h: BRZ algorithm is
	almost stable

2005-07-29 00:09  fc_botelho

	* src/: brz.c, djb2_hash.c, djb2_hash.h, fnv_hash.c, fnv_hash.h,
	hash.c, hash.h, jenkins_hash.c, jenkins_hash.h, sdbm_hash.c,
	sdbm_hash.h: it was fixed more mistakes in BRZ algorithm

2005-07-28 21:00  fc_botelho

	* src/: bmz.c, brz.c, cmph.c: fixed some mistakes in BRZ algorithm

2005-07-27 19:13  fc_botelho

	* src/brz.c: algorithm BRZ included

2005-07-27 18:16  fc_botelho

	* src/: bmz_structs.h, brz.c, brz.h, brz_structs.h: Algorithm BRZ
	included

2005-07-27 18:13  fc_botelho

	* src/: Makefile.am, bmz.c, chm.c, cmph.c, cmph.h, cmph_types.h:
	Algorithm BRZ included

2005-07-25 19:18  fc_botelho

	* README, README.t2t, scpscript: it was included an examples
	directory

2005-07-25 18:26  fc_botelho

	* INSTALL, Makefile.am, configure.ac, examples/Makefile,
	examples/Makefile.am, examples/Makefile.in,
	examples/file_adapter_ex2.c, examples/keys.txt,
	examples/vector_adapter_ex1.c, examples/.deps/file_adapter_ex2.Po,
	examples/.deps/vector_adapter_ex1.Po, src/cmph.c, src/cmph.h: it
	was included a examples directory

2005-03-03 02:07  davi

	* src/: bmz.c, chm.c, chm.h, chm_structs.h, cmph.c, cmph.h,
	graph.c, graph.h, jenkins_hash.c, jenkins_hash.h, main.c (xgraph):
	New f*cking cool algorithm works. Roughly implemented in chm.c

2005-03-02 20:55  davi

	* src/xgraph.c (xgraph): xchmr working nice, but a bit slow

2005-03-02 02:01  davi

	* src/xchmr.h: file xchmr.h was initially added on branch xgraph.

2005-03-02 02:01  davi

	* src/xchmr_structs.h: file xchmr_structs.h was initially added on
	branch xgraph.

2005-03-02 02:01  davi

	* src/xchmr.c: file xchmr.c was initially added on branch xgraph.

2005-03-02 02:01  davi

	* src/: Makefile.am, cmph.c, cmph_types.h, xchmr.c, xchmr.h,
	xchmr_structs.h, xgraph.c, xgraph.h (xgraph): xchmr working fine
	except for false positives on cyclic detection.

2005-03-02 00:05  davi

	* src/: Makefile.am, xgraph.c, xgraph.h (xgraph): Added external
	graph functionality in branch xgraph.

2005-03-02 00:05  davi

	* src/xgraph.c: file xgraph.c was initially added on branch xgraph.

2005-03-02 00:05  davi

	* src/xgraph.h: file xgraph.h was initially added on branch xgraph.

2005-02-28 19:53  davi

	* src/chm.c: Fixed off by one bug in chm.

2005-02-17 16:20  fc_botelho

	* LOGO.html, README, README.t2t, gendocs: The way of calling the
	function cmph_search was fixed in the file README.t2t

2005-01-31 17:13  fc_botelho

	* README.t2t: Heuristic BMZ memory consumption was updated

2005-01-31 17:09  fc_botelho

	* BMZ.t2t: DJB2, SDBM, FNV and Jenkins hash link were added

2005-01-31 16:50  fc_botelho

	* BMZ.t2t, CHM.t2t, COMPARISON.t2t, CONCEPTS.t2t, CONFIG.t2t,
	FAQ.t2t, GPERF.t2t, LOGO.t2t, README.t2t, TABLE1.t2t, TABLE4.t2t,
	TABLE5.t2t, DOC.css: BMZ documentation was finished

2005-01-28 18:12  fc_botelho

	* figs/img1.png, figs/img10.png, figs/img100.png, figs/img101.png,
	figs/img102.png, figs/img103.png, figs/img104.png, figs/img105.png,
	figs/img106.png, figs/img107.png, figs/img108.png, figs/img109.png,
	papers/bmz_tr004_04.ps, papers/bmz_wea2005.ps, papers/chm92.pdf,
	figs/img11.png, figs/img110.png, figs/img111.png, figs/img112.png,
	figs/img113.png, figs/img114.png, figs/img115.png, figs/img116.png,
	figs/img117.png, figs/img118.png, figs/img119.png, figs/img12.png,
	figs/img120.png, figs/img121.png, figs/img122.png, figs/img123.png,
	figs/img124.png, figs/img125.png, figs/img126.png, figs/img127.png,
	figs/img128.png, figs/img129.png, figs/img13.png, figs/img130.png,
	figs/img131.png, figs/img132.png, figs/img133.png, figs/img134.png,
	figs/img135.png, figs/img136.png, figs/img137.png, figs/img138.png,
	figs/img139.png, figs/img14.png, figs/img140.png, figs/img141.png,
	figs/img142.png, figs/img143.png, figs/img144.png, figs/img145.png,
	figs/img146.png, figs/img147.png, figs/img148.png, figs/img149.png,
	figs/img15.png, figs/img150.png, figs/img151.png, figs/img152.png,
	figs/img153.png, figs/img154.png, figs/img155.png, figs/img156.png,
	figs/img157.png, figs/img158.png, figs/img159.png, figs/img16.png,
	figs/img160.png, figs/img161.png, figs/img162.png, figs/img163.png,
	figs/img164.png, figs/img165.png, figs/img166.png, figs/img167.png,
	figs/img168.png, figs/img169.png, figs/img17.png, figs/img170.png,
	figs/img171.png, figs/img172.png, figs/img173.png, figs/img174.png,
	figs/img175.png, figs/img176.png, figs/img177.png, figs/img178.png,
	figs/img179.png, figs/img18.png, figs/img180.png, figs/img181.png,
	figs/img182.png, figs/img183.png, figs/img184.png, figs/img185.png,
	figs/img186.png, figs/img187.png, figs/img188.png, figs/img189.png,
	figs/img19.png, figs/img190.png, figs/img191.png, figs/img192.png,
	figs/img193.png, figs/img194.png, figs/img195.png, figs/img196.png,
	figs/img197.png, figs/img198.png, figs/img199.png, figs/img2.png,
	figs/img20.png, figs/img200.png, figs/img201.png, figs/img202.png,
	figs/img203.png, figs/img204.png, figs/img205.png, figs/img206.png,
	figs/img207.png, figs/img208.png, figs/img209.png, figs/img21.png,
	figs/img210.png, figs/img211.png, figs/img212.png, figs/img213.png,

src/main.c
#ifdef WIN32
#include "wingetopt.h"
#else
#include <getopt.h>
#endif
#include <stdio.h>
#include <stdlib.h>
#include <errno.h>
#include <string.h>
#include <time.h>
#include <limits.h>
#include <assert.h>
#include "cmph.h"
#include "hash.h"

#ifdef WIN32
#define VERSION "0.8"
#else
#include "config.h"
#endif


void usage(const char *prg)
{
	fprintf(stderr, "usage: %s [-v] [-h] [-V] [-k nkeys] [-f hash_function] [-g [-c algorithm_dependent_value][-s seed] ] [-a algorithm] [-M memory_in_MB] [-b algorithm_dependent_value] [-t keys_per_bin] [-d tmp_dir] [-m file.mph]  keysfile\n", prg);
}
void usage_long(const char *prg)
{
	cmph_uint32 i;
	fprintf(stderr, "usage: %s [-v] [-h] [-V] [-k nkeys] [-f hash_function] [-g [-c algorithm_dependent_value][-s seed] ] [-a algorithm] [-M memory_in_MB] [-b algorithm_dependent_value] [-t keys_per_bin] [-d tmp_dir] [-m file.mph] keysfile\n", prg);
	fprintf(stderr, "Minimum perfect hashing tool\n\n");
	fprintf(stderr, "  -h\t print this help message\n");
	fprintf(stderr, "  -c\t c value determines:\n");
	fprintf(stderr, "    \t  * the number of vertices in the graph for the algorithms BMZ and CHM\n");
	fprintf(stderr, "    \t  * the number of bits per key required in the FCH algorithm\n");
	fprintf(stderr, "    \t  * the load factor in the CHD_PH algorithm\n");
	fprintf(stderr, "  -a\t algorithm - valid values are\n");
	for (i = 0; i < CMPH_COUNT; ++i) fprintf(stderr, "    \t  * %s\n", cmph_names[i]);
	fprintf(stderr, "  -f\t hash function (may be used multiple times) - valid values are\n");
	for (i = 0; i < CMPH_HASH_COUNT; ++i) fprintf(stderr, "    \t  * %s\n", cmph_hash_names[i]);
	fprintf(stderr, "  -V\t print version number and exit\n");
	fprintf(stderr, "  -v\t increase verbosity (may be used multiple times)\n");
	fprintf(stderr, "  -k\t number of keys\n");
	fprintf(stderr, "  -g\t generation mode\n");
	fprintf(stderr, "  -s\t random seed\n");
	fprintf(stderr, "  -m\t minimum perfect hash function file \n");
	fprintf(stderr, "  -M\t main memory availability (in MB) used in BRZ algorithm \n");
	fprintf(stderr, "  -d\t temporary directory used in BRZ algorithm \n");
	fprintf(stderr, "  -b\t the meaning of this parameter depends on the algorithm selected in the -a option:\n");
	fprintf(stderr, "    \t  * For BRZ it is used to make the maximal number of keys in a bucket lower than 256.\n");
	fprintf(stderr, "    \t    In this case its value should be an integer in the range [64,175]. Default is 128.\n\n");
	fprintf(stderr, "    \t  * For BDZ it is used to determine the size of some precomputed rank\n");
	fprintf(stderr, "    \t    information and its value should be an integer in the range [3,10]. Default\n");
	fprintf(stderr, "    \t    is 7. The larger is this value, the more compact are the resulting functions\n");
	fprintf(stderr, "    \t    and the slower are them at evaluation time.\n\n");
	fprintf(stderr, "    \t  * For CHD and CHD_PH it is used to set the average number of keys per bucket\n");
	fprintf(stderr, "    \t    and its value should be an integer in the range [1,32]. Default is 4. The\n");
	fprintf(stderr, "    \t    larger is this value, the slower is the construction of the functions.\n");
	fprintf(stderr, "    \t    This parameter has no effect for other algorithms.\n\n");
	fprintf(stderr, "  -t\t set the number of keys per bin for a t-perfect hashing function. A t-perfect\n");
	fprintf(stderr, "    \t hash function allows at most t collisions in a given bin. This parameter applies\n");
	fprintf(stderr, "    \t only to the CHD and CHD_PH algorithms. Its value should be an integer in the\n");
	fprintf(stderr, "    \t range [1,128]. Defaul is 1\n");
	fprintf(stderr, "  keysfile\t line separated file with keys\n");
}

int main(int argc, char **argv)
{
	cmph_uint32 verbosity = 0;
	char generate = 0;
	char *mphf_file = NULL;
	FILE *mphf_fd = stdout;
	const char *keys_file = NULL;
	FILE *keys_fd;
	cmph_uint32 nkeys = UINT_MAX;
	cmph_uint32 seed = UINT_MAX;
	CMPH_HASH *hashes = NULL;
	cmph_uint32 nhashes = 0;
	cmph_uint32 i;
	CMPH_ALGO mph_algo = CMPH_CHM;
	double c = 0;
	cmph_config_t *config = NULL;
	cmph_t *mphf = NULL;
	char * tmp_dir = NULL;
	cmph_io_adapter_t *source;
	cmph_uint32 memory_availability = 0;
	cmph_uint32 b = 0;
	cmph_uint32 keys_per_bin = 1;
	while (1)
	{
		char ch = (char)getopt(argc, argv, "hVvgc:k:a:M:b:t:f:m:d:s:");
		if (ch == -1) break;
		switch (ch)
		{
			case 's':
				{
					char *cptr;
					seed = (cmph_uint32)strtoul(optarg, &cptr, 10);
					if(*cptr != 0) {
						fprintf(stderr, "Invalid seed %s\n", optarg);
						exit(1);
					}
				}
				break;
			case 'c':
				{
					char *endptr;
					c = strtod(optarg, &endptr);
					if(*endptr != 0) {
						fprintf(stderr, "Invalid c value %s\n", optarg);
						exit(1);
					}
				}
				break;
			case 'g':
				generate = 1;
				break;
			case 'k':
			        {
					char *endptr;
					nkeys = (cmph_uint32)strtoul(optarg, &endptr, 10);
					if(*endptr != 0) {
						fprintf(stderr, "Invalid number of keys %s\n", optarg);
						exit(1);
					}
				}
				break;
			case 'm':
				mphf_file = strdup(optarg);
				break;
			case 'd':
				tmp_dir = strdup(optarg);
				break;
			case 'M':
				{
					char *cptr;
					memory_availability = (cmph_uint32)strtoul(optarg, &cptr, 10);
					if(*cptr != 0) {
						fprintf(stderr, "Invalid memory availability %s\n", optarg);
						exit(1);
					}
				}
				break;
			case 'b':
				{
					char *cptr;
					b =  (cmph_uint32)strtoul(optarg, &cptr, 10);
					if(*cptr != 0) {
						fprintf(stderr, "Parameter b was not found: %s\n", optarg);
						exit(1);
					}
				}
				break;
			case 't':
				{
					char *cptr;
					keys_per_bin = (cmph_uint32)strtoul(optarg, &cptr, 10);
					if(*cptr != 0) {
						fprintf(stderr, "Parameter t was not found: %s\n", optarg);
						exit(1);
					}
				}
				break;
			case 'v':
				++verbosity;
				break;
			case 'V':
				printf("%s\n", VERSION);
				return 0;
			case 'h':
				usage_long(argv[0]);
				return 0;
			case 'a':
				{
				char valid = 0;
				for (i = 0; i < CMPH_COUNT; ++i)
				{
					if (strcmp(cmph_names[i], optarg) == 0)
					{
						mph_algo = (CMPH_ALGO)i;
						valid = 1;
						break;
					}
				}
				if (!valid)
				{
					fprintf(stderr, "Invalid mph algorithm: %s. It is not available in version %s\n", optarg, VERSION);
					return -1;
				}
				}
				break;
			case 'f':
				{
				char valid = 0;
				for (i = 0; i < CMPH_HASH_COUNT; ++i)
				{
					if (strcmp(cmph_hash_names[i], optarg) == 0)
					{
						hashes = (CMPH_HASH *)realloc(hashes, sizeof(CMPH_HASH) * ( nhashes + 2 ));
						hashes[nhashes] = (CMPH_HASH)i;
						hashes[nhashes + 1] = CMPH_HASH_COUNT;
						++nhashes;
						valid = 1;
						break;
					}
				}
				if (!valid)
				{
					fprintf(stderr, "Invalid hash function: %s\n", optarg);
					return -1;
				}
				}
				break;
			default:
				usage(argv[0]);
				return 1;
		}
	}

	if (optind != argc - 1)
	{
		usage(argv[0]);
		return 1;
	}
	keys_file = argv[optind];

	if (seed == UINT_MAX) seed = (cmph_uint32)time(NULL);
	srand(seed);
	int ret = 0;
	if (mphf_file == NULL)
	{
		mphf_file = (char *)malloc(strlen(keys_file) + 5);
		memcpy(mphf_file, keys_file, strlen(keys_file));
		memcpy(mphf_file + strlen(keys_file), ".mph\0", (size_t)5);
	}

	keys_fd = fopen(keys_file, "r");

	if (keys_fd == NULL)
	{
		fprintf(stderr, "Unable to open file %s: %s\n", keys_file, strerror(errno));
		return -1;
	}

	if (seed == UINT_MAX) seed = (cmph_uint32)time(NULL);
	if(nkeys == UINT_MAX) source = cmph_io_nlfile_adapter(keys_fd);
	else source = cmph_io_nlnkfile_adapter(keys_fd, nkeys);
	if (generate)
	{
		//Create mphf
		mphf_fd = fopen(mphf_file, "w");
		config = cmph_config_new(source);
		cmph_config_set_algo(config, mph_algo);
		if (nhashes) cmph_config_set_hashfuncs(config, hashes);
		cmph_config_set_verbosity(config, verbosity);
		cmph_config_set_tmp_dir(config, (cmph_uint8 *) tmp_dir);
		cmph_config_set_mphf_fd(config, mphf_fd);
		cmph_config_set_memory_availability(config, memory_availability);
		cmph_config_set_b(config, b);
		cmph_config_set_keys_per_bin(config, keys_per_bin);

		//if((mph_algo == CMPH_BMZ || mph_algo == CMPH_BRZ) && c >= 2.0) c=1.15;
		if(mph_algo == CMPH_BMZ  && c >= 2.0) c=1.15;
		if (c != 0) cmph_config_set_graphsize(config, c);
		mphf = cmph_new(config);

		cmph_config_destroy(config);
		if (mphf == NULL)
		{
			fprintf(stderr, "Unable to create minimum perfect hashing function\n");
			//cmph_config_destroy(config);
			free(mphf_file);
			return -1;
		}

		if (mphf_fd == NULL)
		{
			fprintf(stderr, "Unable to open output file %s: %s\n", mphf_file, strerror(errno));
			free(mphf_file);
			return -1;
		}
		cmph_dump(mphf, mphf_fd);
		cmph_destroy(mphf);
		fclose(mphf_fd);
	}
	else
	{
		cmph_uint8 * hashtable = NULL;
		mphf_fd = fopen(mphf_file, "r");
		if (mphf_fd == NULL)
		{
			fprintf(stderr, "Unable to open input file %s: %s\n", mphf_file, strerror(errno));
			free(mphf_file);
			return -1;
		}
		mphf = cmph_load(mphf_fd);
		fclose(mphf_fd);
		if (!mphf)
		{
			fprintf(stderr, "Unable to parser input file %s\n", mphf_file);
			free(mphf_file);
			return -1;
		}
		cmph_uint32 siz = cmph_size(mphf);
		hashtable = (cmph_uint8*)calloc(siz, sizeof(cmph_uint8));
		memset(hashtable, 0,(size_t) siz);
		//check all keys
		for (i = 0; i < source->nkeys; ++i)
		{
			cmph_uint32 h;
			char *buf;
			cmph_uint32 buflen = 0;
			source->read(source->data, &buf, &buflen);
			h = cmph_search(mphf, buf, buflen);
			if (!(h < siz))
			{
				fprintf(stderr, "Unknown key %*s in the input.\n", buflen, buf);
				ret = 1;
			} else if(hashtable[h] >= keys_per_bin)
			{
				fprintf(stderr, "More than %u keys were mapped to bin %u\n", keys_per_bin, h);
				fprintf(stderr, "Duplicated or unknown key %*s in the input\n", buflen, buf);
				ret = 1;
			} else hashtable[h]++;

			if (verbosity)
			{
				printf("%s -> %u\n", buf, h);
			}
			source->dispose(source->data, buf, buflen);
		}

		cmph_destroy(mphf);
		free(hashtable);
	}
	fclose(keys_fd);
	free(mphf_file);
	free(tmp_dir);
        cmph_io_nlfile_adapter_destroy(source);
	return ret;

}
